flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/LegacyUserDefinedFunctionInference.java

```diff
@@ -0,0 +1,529 @@
+        if (accType != null) {
+            return fromLegacyInfoToDataType(accType);
+        } else {
+            try {
+                return fromLegacyInfoToDataType(
+                        TypeExtractor.createTypeInfo(
+                                func, ImperativeAggregateFunction.class, func.getClass(), 1));
+            } catch (InvalidTypesException ite) {
+                throw new TableException(
+                        String.format(
+                                "Cannot infer generic type of %s}. You can override"
+                                        + " ImperativeAggregateFunction.getAccumulatorType()"
+                                        + " to specify the type.",
+                                func.getClass()),
+                        ite);
+            }
+        }
+    }
+
+    private static Class<?>[] logicalTypesToExternalClasses(LogicalType[] types) {
+        return Arrays.stream(types)
+                .map(
+                        t -> {
+                            if (t == null) {
+                                return null;
+                            } else {
+                                return TypeConversions.fromLogicalToDataType(t)
+                                        .getConversionClass();
+                            }
+                        })
+                .toArray(Class<?>[]::new);
+    }
+
+    private static Class<?>[] logicalTypesToInternalClasses(LogicalType[] types) {
+        return Arrays.stream(types)
+                .map(
+                        t -> {
+                            if (t == null) {
+                                return null;
+                            } else {
+                                return toInternalConversionClass(t);
+                            }
+                        })
+                .toArray(Class<?>[]::new);
+    }
+
+    private static boolean parameterClassEquals(Class<?> candidate, Class<?> expected) {
+        return candidate == null
+                || candidate == expected
+                || expected == Object.class
+                || candidate == Object.class
+                || // Special case when we don't know the type
+                expected.isPrimitive() && Primitives.wrap(expected) == candidate
+                || candidate == Date.class && (expected == Integer.class || expected == int.class)
+                || candidate == Time.class && (expected == Integer.class || expected == int.class)
+                || candidate == StringData.class && expected == String.class
+                || candidate == String.class && expected == StringData.class
+                || candidate == TimestampData.class && expected == LocalDateTime.class
+                || candidate == Timestamp.class && expected == TimestampData.class
+                || candidate == TimestampData.class && expected == Timestamp.class
+                || candidate == LocalDateTime.class && expected == TimestampData.class
```

Comment:
(nit) Reorder:
```
                || candidate == TimestampData.class && expected == LocalDateTime.class
                || candidate == LocalDateTime.class && expected == TimestampData.class
                || candidate == Timestamp.class && expected == TimestampData.class
                || candidate == TimestampData.class && expected == Timestamp.class
```

Reply:
I'd rather keep the original order from `UserDefinedFunctionUtils`